pkg/cloud-controller-manager/loadbalancer.go

```diff
@@ -0,0 +1,371 @@
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: lbv1.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: l.namespace,
+				Name:      name,
+			},
+		}
+	} else {
+		lb = oldLB.DeepCopy()
+	}
+
+	if lb.Annotations == nil {
+		lb.Annotations = make(map[string]string)
+	}
+	lb.Annotations[pkgctllb.AnnotationKeyNetwork] = service.Annotations[KeyNetwork]
+	lb.Annotations[pkgctllb.AnnotationKeyProject] = service.Annotations[KeyProject]
+	lb.Annotations[pkgctllb.AnnotationKeyNamespace] = service.Annotations[KeyNamespace]
+	lb.Annotations[pkgctllb.AnnotationKeyCluster] = clusterName
+
+	if lb.Labels == nil {
+		lb.Labels = make(map[string]string)
+	}
+	lb.Labels[clusterNameKey] = clusterName
+	lb.Labels[serviceNamespaceKey] = service.Namespace
+	lb.Labels[serviceNameKey] = service.Name
+
+	ipam := lbv1.Pool
+	if ipamStr, ok := service.Annotations[KeyIPAM]; ok {
+		ipam = lbv1.IPAM(ipamStr)
+	}
+	lb.Spec.IPAM = ipam
+	lb.Spec.WorkloadType = lbv1.Cluster
+
+	if len(service.Status.LoadBalancer.Ingress) > 0 {
+		lb.Status.Address = service.Status.LoadBalancer.Ingress[0].IP
+	}
+
+	return lb
+}
+
+func (l *LoadBalancerManager) updateServiceLoadBalancerIP(lbName string, service *v1.Service) error {
+	object, ip, err := waitForIP(func() (runtime.Object, string, error) {
+		lb, err := l.lbClient.Get(l.namespace, lbName, metav1.GetOptions{})
+		if err != nil || lb.Status.AllocatedAddress.IP == "" {
+			return nil, "", fmt.Errorf("could not get allocated IP address")
+		}
+		return lb, lb.Status.AllocatedAddress.IP, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	lb := object.(*lbv1.LoadBalancer)
+	if service.Annotations != nil && service.Annotations[KeyKubevipLoadBalancerIP] == ip && lb.Status.Address == ip {
+		return nil
+	}
+
+	serviceCopy := service.DeepCopy()
+	if serviceCopy.Annotations == nil {
+		serviceCopy.Annotations = make(map[string]string)
+	}
+	serviceCopy.Annotations[KeyKubevipLoadBalancerIP] = ip
+	if _, err := l.localSvcClient.Update(serviceCopy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LoadBalancerManager) updateSecondaryServiceLoadBalancerIP(ip string, service *v1.Service) error {
+	if service.Annotations != nil && service.Annotations[KeyKubevipLoadBalancerIP] == ip && service.Annotations[KeyIPAM] == "" {
+		return nil
+	}
+
+	serviceCopy := service.DeepCopy()
+	if serviceCopy.Annotations == nil {
+		serviceCopy.Annotations = make(map[string]string)
+	}
+	// update the annotations and kube-vip will update the service status load balancer
+	serviceCopy.Annotations[KeyKubevipLoadBalancerIP] = ip
+	delete(serviceCopy.Annotations, KeyIPAM)
+	if _, err := l.localSvcClient.Update(serviceCopy); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LoadBalancerManager) deleteLoadBalancer(clusterName string, service *v1.Service) error {
+	// check if there are other services using the same load balancer
+	if err := l.checkSecondaryServicesBeforeDeleted(service); err != nil {
```

Comment:
you may also need to check if the secondary svc exists before updating a primary svc as secondary.

Reply:
In the process of ensuring secondary service, it will delete load balancer if existing. Therefore, we don't have check again.